spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Scheduler.scala

```diff
@@ -0,0 +1,308 @@
+ *
+ * @param inputNodes start nodes
+ * @param outputNodes target nodes
+ * @tparam T
+ */
+private[bigdl] class Scheduler[T](
+    inputNodes: Seq[ModuleNode[T]], outputNodes: Seq[ModuleNode[T]]
+  ) {
+
+  import Scheduler._
+
+  private val readyQueue = new mutable.Queue[ModuleNode[T]]()
+  private val nodeStatus = new NodeStatusManager[T]()
+
+  /**
+   * User must reset the scheduler after first use it or finish a graph execution
+   */
+  def reset(): Unit = {
+    readyQueue.clear()
+    inputNodes.foreach(n => {
+      readyQueue.enqueue(n)
+    })
+    nodeStatus.removeUnConstStatus()
+  }
+
+
+  /**
+   * If every output nodes is executed. Please note if some of the output nodes has not been
+   * executed but the execution can't move forward, an exception will be thrown
+   * @return
+   */
+  def isFinished(): Boolean = {
+    val isEmpty = readyQueue.isEmpty
+    if (isEmpty) {
+      outputNodes.foreach(n => {
+        require(!nodeStatus.notExecuted(n), "Some output nodes have not been executed")
+      })
+    }
+    isEmpty
+  }
+
+
+  /**
+   * Fetch a node to execute. Don't call it when isFinished is true, or it will throw an exception
+   * @return
+   */
+  def fetch(): ModuleNode[T] = {
+    var node = readyQueue.dequeue()
+    while (nodeStatus.isConst(node)) {
+      schedule(node)
+      node = readyQueue.dequeue()
+    }
+    node
+  }
+
+  /**
+   * Schedule nodes depend on the given node
+   * @param node
+   */
+  def schedule(node: ModuleNode[T]): Unit = {
+    // Update status of current node
+    node.element match {
+      case e: Enter[_] =>
+        require(node.prevNodes.length == 1, "Enter only accept one node input")
+        nodeStatus(node) = nodeStatus.stackIteration(nodeStatus(node.prevNodes.head))
+      case x: Exit[_] =>
+        require(node.prevNodes.length == 1, "Exit only accept one node input")
+        nodeStatus(node) = nodeStatus.unstackIteration(nodeStatus.iteration(node.prevNodes.head))
+      case n: NextIteration[_] =>
+        require(node.prevNodes.length == 1, "NextIteration only accept one node input")
+        val lastIteration = nodeStatus.iteration(node.prevNodes.head)
+        lastIteration.outOfDate = true
+        nodeStatus(node) = Iteration(lastIteration.prev)
+      case _ =>
+        nodeStatus(node) = if (node.prevNodes.length == 0) {
+          if (node.element.isInstanceOf[com.intel.analytics.bigdl.nn.tf.Const[_]]) {
+            Const()
+          } else {
+            Ready()
+          }
+        } else {
+          val constNodes = node.prevNodes.filter(nodeStatus.isConst(_))
+          if (constNodes.length == node.prevNodes.length) {
+            Const()
+          } else {
+            val iterationNodes = node.prevNodes.filter(nodeStatus.isIteration(_))
+              .filter(!nodeStatus.iteration(_).outOfDate)
+              .filter(!_.element.isInstanceOf[LoopCondition[_]])
```

Comment:
Is it possible that two different loop variables have some kind interaction, so that the iteration of one loop variable can be falsely passed to the other one by some connection other than the "LoopCondition" node